pkgs/build-support/singularity-tools/default.nix

```diff
@@ -108,14 +116,14 @@ rec {
 
             # Build /bin and copy across closure
             mkdir -p bin ./${builtins.storeDir}
-            for f in $(cat $layerClosure) ; do
-              cp -ar $f ./$f
-            done
+            while IFS= read -r f; do
+              cp -r "$f" "./$f"
+            done < "$layerClosure"
 
-            for c in ${toString contents} ; do
-              for f in $c/bin/* ; do
-                if [ ! -e bin/$(basename $f) ] ; then
-                  ln -s $f bin/
+            for c in ${lib.escapeShellArgs contents} ; do
```

Comment:
Some people would disagree, but I think that passing values as bash variables or files is preferable to template rendering, especially when it comes to this kind of potentially long lists.

That is, I'd `read` from `$contents`, or, better yet, enable `__structuredAttrs` and iterate over `"${contents[@]}"`

Reply:
I agree that storing a long list of items in a variable and passing it to the Bash string is more elegant than dumping a long line into it. Unfortunately, `vmTools.runInLinuxVM` currently fails when the input derivation has `__structuredAttrs = true` (#334705).

Even though `contents` would most likely be out paths of packages, which are unlikely to contain spaces, I still hope to prevent unexpected splits of a space-containing path. Passing an attribute that concatenates `contents` with `\n` or `\0` just to read from it might confuse future developers.

Reply:
I'll add a TODO comment then.

Reply:
Addressed.